rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,366 @@
+            };
+
+            let y = rng.sample(Standard);
+            if y < ratio {
+                return x;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn test_samples<F: Float + core::fmt::Debug, D: Distribution<F>>(
+        distr: D, zero: F, expected: &[F],
+    ) {
+        let mut rng = crate::test::rng(213);
+        let mut buf = [zero; 4];
+        for x in &mut buf {
+            *x = rng.sample(&distr);
+        }
+        assert_eq!(buf, expected);
+    }
+
+    #[test]
+    #[should_panic]
+    fn zeta_invalid() {
+        Zeta::new(1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zeta_nan() {
+        Zeta::new(core::f64::NAN).unwrap();
+    }
+
+    #[test]
+    fn zeta_sample() {
+        let a = 2.0;
+        let d = Zeta::new(a).unwrap();
+        let mut rng = crate::test::rng(1);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zeta_small_a() {
+        let a = 1. + 1e-15;
+        let d = Zeta::new(a).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zeta_value_stability() {
+        test_samples(Zeta::new(1.5).unwrap(), 0f32, &[
+            1.0, 2.0, 1.0, 1.0,
+        ]);
+        test_samples(Zeta::new(2.0).unwrap(), 0f64, &[
+            2.0, 1.0, 1.0, 1.0,
+        ]);
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_s_too_small() {
+        Zipf::new(10, -1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_n_too_small() {
+        Zipf::new(0, 1.).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn zipf_nan() {
+        Zipf::new(10, core::f64::NAN).unwrap();
+    }
+
+    #[test]
+    fn zipf_sample() {
+        let d = Zipf::new(10, 0.5).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zipf_sample_s_1() {
+        let d = Zipf::new(10, 1.).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+    }
+
+    #[test]
+    fn zipf_sample_s_0() {
+        let d = Zipf::new(10, 0.).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+        // TODO: verify that this is a uniform distribution
+    }
+
+    #[test]
+    fn zipf_sample_large_n() {
+        let d = Zipf::new(core::u64::MAX, 1.5).unwrap();
+        let mut rng = crate::test::rng(2);
+        for _ in 0..1000 {
+            let r = d.sample(&mut rng);
+            assert!(r >= 1.);
+        }
+        // TODO: verify that this is a zeta distribution
```

Comment:
To verify this we would the zeta function (to compute the distribution of a Zeta variable).
Some implementations are given in:
- [rgsl::zeta::riemann::zeta](https://docs.rs/GSL/4.0.2/rgsl/zeta/riemann/fn.zeta.html)
- [spfunc::zeta::zeta](https://docs.rs/spfunc/0.1.0/spfunc/zeta/fn.zeta.html)

And I saw that the crate `average` was added in `dev-dependencies` for testing purposes, so maybe there is space for more(?) In this case, maybe `rgsl` is a better option, as it should be helpful in other tests too.

Reply:
I would prefer a Rust implementation over GSL bindings for portability reasons, so `spfunc` looks like a better option to me.

Reply:
Okay!! If we need more functions, we could contribute to `spfunc` in the future :)

(I still can not close the comment, so please do it)